src/Language/PureScript/TypeChecker.hs

```diff
@@ -712,3 +713,19 @@ typeCheckModule (Module ss coms mn decls (Just exps)) =
     extractMemberName (TypeDeclaration td) = tydeclIdent td
     extractMemberName _ = internalError "Unexpected declaration in typeclass member list"
   checkClassMembersAreExported _ = return ()
+
+  checkDataConstructorsAreExported :: DeclarationRef -> m ()
+  checkDataConstructorsAreExported dr@(TypeRef ss' name (Just exportedDataConstructorsNames))
+    | not (null exportedDataConstructorsNames) = do
+      let dataConstructorNames = fromMaybe [] . headMay $ mapMaybe findDataConstructorsNames decls
```

Comment:
Data constructors are all in the environment, so we shouldn't need to filter through all declarations for every data export. They are under `types`, and part of the unfortunately named `TypeKind` slot.